include/small/util.h

```diff
@@ -58,3 +113,271 @@ small_getpagesize(void)
+ */
+static inline size_t
+small_align(size_t size, size_t alignment)
+{
+	/* Must be a power of two */
+	assert((alignment & (alignment - 1)) == 0);
+	/* Bit arithmetics won't work for a large size */
+	assert(size <= SIZE_MAX - alignment);
+	return (size - 1 + alignment) & ~(alignment - 1);
+}
+
+/**
+ * Align value to the nearest divisible by the given alignment which is
+ * not greater than value. Alignment must be a power of 2.
+ */
+static inline size_t
+small_align_down(size_t value, size_t alignment)
+{
+	assert((alignment & (alignment - 1)) == 0);
+	return value & ~(alignment - 1);
+}
+
+/** Round up a number to the nearest power of two. */
+static inline size_t
+small_round(size_t size)
+{
+	if (size < 2)
+		return size;
+	assert(size <= SIZE_MAX / 2 + 1);
+	assert(size - 1 <= ULONG_MAX);
+	size_t r = 1;
+	return r << (sizeof(unsigned long) * CHAR_BIT -
+		     __builtin_clzl((unsigned long) (size - 1)));
+}
+
+/** Binary logarithm of a size. */
+static inline size_t
+small_lb(size_t size)
+{
+	assert(size <= ULONG_MAX);
+	return sizeof(unsigned long) * CHAR_BIT -
+		__builtin_clzl((unsigned long) size) - 1;
+}
+
+#ifdef SMALL_MALLOC_IMPL
+
+/**
+ * Small wrapper allows to hold extra information (header) for every user
+ * memory allocation. For this purpose we allocate more memory then requested
+ * and keep a header and user data in a same chunk of memory. So wrapper memory
+ * layout is next:
+ *
+ *     UUU HHHHHHH AAA PPPPPPPP UUUU
+ *
+ * Where
+ *    P - user payload
+ *    H - header
+ *    A - memory used for alignment
+ *    F - unused memory (raised to header and payload alignments)
+ *
+ * Header has SMALL_HEADER_ALIGNMENT (large alignment to allow any
+ * types in header). User payload is aligned to given alignment A.
+ * Additionally payload is not aligned on 2 * A. This improves unaligned
+ * access check. Header should be inherited from struct small_header.
+ *
+ * Wrapper is used with ASAN builds. Thus to check buffer underflows/overflows
+ * wrapper memory is poisoned except for user payload of course. Note that
+ * due to poison alignment restrictions we may not be able to poison
+ * part of alignment area before payload. In this case we write magic to
+ * the area which cannot be poisoned and check it is not changed when memory
+ * is freed.
+ */
+
+enum {
+	/* Alignment of header in wrapper. */
+	SMALL_HEADER_ALIGNMENT = sizeof(long long),
+	/*
+	 * ASAN does not allow to precisely poison arbitrary ranges of memory.
+	 * However if range end is 8-aligned or range end is end of memory
+	 * allocated with malloc() then poison is precise.
+	 */
+	SMALL_POISON_ALIGNMENT = 8,
+};
+
+/** Random magic to be used for memory that cannot be poisoned. */
+extern uint64_t small_magic;
+
+/** Wrapper header base. Should be used as first member of actual header. */
+struct small_header {
+	/** Distance from header to wrapper begin in bytes. */
+	size_t alloc_offset;
+};
+
+/**
+ * Wrapper pointers.
+ */
+struct small_wrapper {
```

Comment:
The "wrapper" concept looks confusing to me. I'd prefer an API like this:

```C
/* returns payload and header in the out-argument */
void *
small_asan_alloc_with_header(size_t payload_size, size_t payload_align,
			     size_t header_size, void **header);

/* returns payload */
void *
small_asan_alloc(size_t payload_size, size_t payload_align);

void
small_asan_free(void *payload);
```

That is,
- no inheritance of a header used by a concrete allocator from the base header; the base header is opaque;
- clear api: `alloc` returns a pointer that is freed by `free`; no struts initialized on stack.

Also, I think we should add "asan" to names to emphasize that this is required for ASAN checks.

Probably, worth discussing with @alyapunov.

Reply:
API after f2f discussion:
```c
/**
 * Allocate aligned memory (payload) with extra place for header.
 *
 * Header has SMALL_ASAN_HEADER_ALIGNMENT (large enough alignment for use with
 * existing small allocators). Payload is aligned to `alignment`,
 * additionally payload is not aligned on 2 * `alignment`. This improves
 * unaligned access check.
 *
 * All allocated memory except for payload is poisoned. Note that due to poison
 * alignment restrictions we may not be able to poison part of alignment area
 * before payload. In this case we write magic to the area which cannot be
 * poisoned and check it is not changed when memory is freed.
 *
 * Return pointer to header. small_asan_header_from_payload
 */
void *
small_asan_alloc(size_t payload_size, size_t alignment, size_t header_size);

/**
 * Return pointer to payload given pointer to header allocated with
 * small_asan_alloc.
 */
void *
small_asan_payload_from_header(void *header);

/**
 * Return pointer to header given pointer to payload (for allocations
 * done thru small_asan_alloc).
 */
void *
small_asan_header_from_payload(void *payload);

/** Free memory allocated with small_asan_alloc. Takes pointer to header. */
void
small_asan_free(void *header);
```